internal/policies/mount/mount.go

```diff
@@ -0,0 +1,215 @@
+
+// Manager holds information needed for handling the mount policies.
+type Manager struct {
+	mountsMu map[string]*sync.Mutex
+	muMu     sync.Mutex
+
+	runDir string
+
+	userLookup   func(string) (*user.User, error)
+	systemCtlCmd []string
+}
+
+// New creates a Manager to handle mount policies.
+func New(runDir string, opts ...Option) (m *Manager, err error) {
+	defer decorate.OnError(&err, i18n.G("failed to create new mount manager"))
+	o := options{
+		userLookup: user.Lookup,
+	}
+
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	// Multiple users will be in users/ subdirectory. Create the main one.
+	// #nosec G301 - multiple users will be in users/ subdirectory, we want all of them to be able to access its own subdirectory.
+	if err := os.MkdirAll(filepath.Join(runDir, "users"), 0750); err != nil {
+		return nil, err
+	}
+
+	return &Manager{
+		mountsMu: make(map[string]*sync.Mutex),
+
+		runDir:       runDir,
+		userLookup:   o.userLookup,
+		systemCtlCmd: o.systemctlCmd,
+	}, nil
+}
+
+// ApplyPolicy generates mount policies based on a list of entries.
+func (m *Manager) ApplyPolicy(ctx context.Context, objectName string, isComputer bool, entries []entry.Entry) (err error) {
+	defer decorate.OnError(&err, i18n.G("can't apply mount policy to %s"), objectName)
+
+	if len(entries) == 0 {
+		return nil
+	}
+
+	log.Debugf(ctx, "Applying mount policy to %s", objectName)
+
+	// Mutexes are per user1, user2, computer
+	m.muMu.Lock()
+	if _, exists := m.mountsMu[objectName]; !exists {
+		m.mountsMu[objectName] = &sync.Mutex{}
+	}
+	m.muMu.Unlock()
+	m.mountsMu[objectName].Lock()
+	defer m.mountsMu[objectName].Unlock()
+
+	for _, entry := range entries {
+		switch entry.Key {
+		case "user-mounts":
+			if isComputer {
+				break
+			}
+
+			if e := m.applyUserPolicy(ctx, objectName, entry); e != nil {
+				err = e
+			}
+
+		default:
+			log.Debugf(ctx, "Key %q is currently not supported by the mount manager", entry.Key)
+		}
+	}
+
+	return err
+}
+
+func (m *Manager) applyUserPolicy(ctx context.Context, username string, entry entry.Entry) (err error) {
+	defer decorate.OnError(&err, i18n.G("failed to apply policy for user %q"), username)
+
+	log.Debugf(ctx, "Applying mount policy to user %q", username)
+
+	var uid, gid int
+	usr, err := m.userLookup(username)
+	if err != nil {
+		return fmt.Errorf(i18n.G("could not retrieve user for %q: %w"), err)
+	}
+	if uid, err = strconv.Atoi(usr.Uid); err != nil {
+		return fmt.Errorf(i18n.G("couldn't convert %q to a valid uid for %q"), usr.Uid, username)
+	}
+	if gid, err = strconv.Atoi(usr.Gid); err != nil {
+		return fmt.Errorf(i18n.G("couldn't convert %q to a valid gid for %q"), usr.Gid, username)
+	}
+
+	objectPath := filepath.Join(m.runDir, "users", usr.Uid)
+	mountsPath := filepath.Join(objectPath, "mounts")
+
+	// Removes the current mounts file, if it exists, before continuing applying the policy.
+	if err = os.Remove(mountsPath); err != nil && !errors.Is(err, os.ErrNotExist) {
```

Comment:
Here, we create a race with no file available (I know, we have it in the script directory too, but let’s not have it here as it’s a single file).

Remember that we rename the file from .new to final name. My suggestion then is to only remove the file if: there is no entry. Btw, it seems that it’s not the case anymore as you won’t enter this function if there are no more entries. -> I suggest, in addition to the fix, to create a new test then (`pathAlreadyExists=true`, apply empty entries -> golden shuld have no file)

For instance, here, you remove the file, a parsing fails -> no more mount file (not even the previous one).

Reply:
Ack on the first part, it seems though that the additional test case is not added yet?

Reply:
It is, isn't it?

Reply:
Hum, I don’t see a test case with a pre-existing mounts file, which is then emptied/removed on ApplyPolicy call? What test case are you talking about?

Reply:
The test case "mount file is removed on refreshing policy with no entries" falls under that category if I'm not mistaken.

Reply:
ah, you did it with a second call, this is fine then! My last remaining part is checking the golden files, but this is just to double check on what you rereviewed :)